cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+
+### ARM Deployments
+
+When using Bicep or ARM, deployments happen by creating an ARM Deployment object, which has a name that makes up part of it's identity. In addition, when preforming a subscription level deployment (the only form `azd` currently supports) you need to provide a location for where the deployment metadata is stored. When we support targeting resource groups, instead of a location we'll need a resource group name.
+
+For now, we'll introduce a `target` property in the environment:
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "target": {
+        "type": "arm",
+        "subscriptionId": "faa080af-c1d8-40ad-9cce-e1a450ca5b57",
+        "location": "westus2"
+      }
+    }
+  }
+}
+```
+
+We should have some way to manage the target (`infra target config`?) and at a minimum it should be easy to understand what subscription or subscription/resource group will be used for deployment (`infra target show`?) For now, these values will be prompted for during the initial deployment (just like other required deployment parameters). The required settings to configure a target will not be uniform across infrastructure providers.  When we introduce resource group level deployments we will do by adding the resource group name as a required piece of target configuration (and remove location).
+
+As a UX concern, we may need to handle the "location" parameter in a clever way to reuse whatever location is being set for the template. Ideally we will not have to ask separate questions for "where do you want to store your resources" and "where do you want to store your deployment metadata" even though these are individually configurable.
+
+Note that the target **does not** contain information about a deployment name. Today the name we use for the deployment matches the environment name. We will change this to instead include the *project name* (stored in the name property of `azure.yaml` and the *environment id* to ensure uniqueness. We should also include some per deployment entropy (or perhaps just a timestamp) such that each deployment gets its own deployment object. This should help with some problems we face today where re-using the same deployment object can confuse our status reporting logic, and provides a better view of deployment history.  
+
+Since deployment names are limited to 64 characters, we will need someway to truncate names if we hit this limit.  We should bias towards keeping as much entropy as possible in the strings to ensure uniqueness at a given scope. We should tag these deployment objects with the full environment ID for querying later.
+
+## User Impacting Changes
```

Comment:
The bicep parameter files with the resulting values should be outputted to the env directory so the user can then later use them without azd if they'd like to.  And they can use it as a debugging tool to see what the values are being passed.